### ISSUE TITLE
Support `--no-checksum` on a package-by-package basis

### ISSUE THREAD

**Issue by bartlettroscoe** [LLNL/spack#29837]
### Summary

I would be great to extend Spack to support `--no-checksum` on a package-by-package basis so that users could remove or add checksum checking on specific packages.  One idea would be to use a `no-checksum` variant like:

```
$ spack install [a lot of other specs] cmake@3.23.0+no-checksum
```

such as with a release of Spack (e.g. v0.17.1) that does not list the version `3.23.0` in the `cmake/package.py` file.

(Other interface knobs are possible as well).

That way, all of the packages would assert the checksum of what they download and just that specific package could not assert the checksum for a version that is not known to that release of Spack.  (In this case, we trust version tags downloaded form the github.com/kitware/cmake project where we may not trust other sites we use to download packages from.)

### Rationale

Currently , as I was just made aware of in https://github.com/trilinos/Trilinos/issues/10355#issuecomment-1084910094, Spack supports installing versions of a package newer than existed when a Spack release came out.  (In this case CMake 3.22 and 3.23 came out after the last Spack v0.17.1 was tagged.)  But to install a newer version of CMake with Spack 0.17.1 without a runtime challenge question, you have to pass in the option `--no-checksum`.  So if you wanted to install a bunch of different packages for which Spack had checksum info for, in addition to installing a newer CMake version for which no checksum info exists, and to do so in one `spack install` invocation, you would have to run:

```
$ spack install --no-checksum [a bunch of package specs] cmake@3.23
``` 

But the problem with this is that you would have to skip checking the checksums for all of the other packages.  That greatly reduces the safety of doing a Spack install.

To avoid this, one could run the install in two invocations of `spack install` as:

```
$ spack install [a bunch of package specs]
$ spack install --no-checksum cmake@3.23
```

That would apply checksum checks for all of the packages except `cmake@3.23`.  However, for teams that are trying to use a single `spack install` invocation, that breaks their process.

Therefore, to maintain a single `spack install` invocation, one could disable the checksum check for just the one `cmake@3.23` package using a targeted `no-checksum` variant as:

```
$ spack install [a bunch of package specs] cmake@3.23+no-checksum
```

Now, that uses checksum checks for every packages **except** `cmake@3.23`.

### Description

Have Spack automatically support the `no-checksum` variant for all packages where it makes sense (i.e. that allow the download of arbitrary versions where the `<package>/package.py` file may not have checksum info).

### Additional information

I searched the must up-to-date development documentation for Spack for [checksum](https://spack.readthedocs.io/en/latest/search.html?q=checksum&check_keywords=yes&area=default#) and could not find any hits for a `checksum` variant. I also searched all of the [Spack issues that have 'checksum' in the Issue title](https://github.com/spack/spack/issues?q=is%3Aissue+in%3Atitle+checksum+) and could not find an existing issue that mentions this idea.  

### General information

- [X] I have run `spack --version` and reported the version of Spack
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by bartlettroscoe** [comment]
NOTE: Another user had the same question/request on March 28 in Slack:

* https://spackpm.slack.com/archives/C5VL7V81G/p1648471813142149

**Comment by adamjstewart** [comment]
I agree that a flag that only affects a specific package would be useful, but this should not be done with variants. Variants affect the build hash, meaning that if you install `cmake+no-checksum`, then build something that depends on `cmake`, it will start building `cmake~no-checksum` even though that exact installation already exists.

**Comment by alalazo** [comment]
Closing this issue since a variant, as explained by @adamjstewart, is not the right way to do this since it affects the hash. Also, two lines in a script doesn't seem to me much worse than a single line (but that is just my opinion).

**Comment by bartlettroscoe** [comment]
> I agree that a flag that only affects a specific package would be useful, but this should not be done with variants. 

Actually, I might disagree with this.  Having a different hash for a package that is installed with checksum checking and without checksum checking may not be a bad thing.  This would allow for an initial install with `--no-checksum` and then a reinstall with a later version of Spack where that version of cmake/package.py file is updated to list the CMake version and its checksum and users and downstream packages might like to be able to differentiate those installs.  Given the recent focus on security, that might be a good thing.

> I agree that a flag that only affects a specific package would be useful

Then let's broaden the scope of this issue to be  "Support turning checksum checking off on a package-by-package basis in a single spack install invocation" for that and reopen this Issue?   If you can extend the Spack command-line and YAML input file specification to support this, then great.  As a user I don't care how it gets supported, just that it does.

> Also, two lines in a script doesn't seem to me much worse than a single line (but that is just my opinion).

Personally I agree (and my prototype for a Spack installation process used at least 4 different invocations of `spack install`) but there are some groups that are demanding a single `spack install` command.  However, I can point to this closed issue (and that Spack does not support package-specific tweaks like this) as a motivation for them to script their installations with Spack to use a number of staged calls to `spack install` instead of just one `spack install` invocation.
